README
mailpie - e-mail full text search

PURPOSE

mailpie is a suite of programs for commandline full-text search of large
e-mail archives.  Keep your inbox uncluttered while retaining the
ability to quickly find an old message that becomes relevant again.

In fact, mailpie performs searches much more quickly than many MUAs'
built in search facilities because it uses a time-tested full text
indexer (swish-e).


REQUIREMENTS

- python with bsddb support (tested with python 2.4.3 and 2.5.2)
- swish-e (tested with swish-e 2.4.3 and 2.4.5)

mailpie is developed on Linux systems.


INSTALLATION

Site-wide installation:
    $ sudo python setup.py install

Personal installation:
    $ python setup.py install --home=$HOME
Make sure that $HOME/lib/python is on PYTHONPATH

See distutils documentation for more installation options.


USAGE

To add a mailbox full of messages to the mailpie storage:
    mailpie-add example.mbox
After adding messages with mailpie-add, the original mbox file is not
needed to perform searches.

To search for messages:
    mailpie-search --after="April 1" from=jepler mutt
This will find messages dated after April 1 where the From: line matches
'jepler' and the header or body matches 'mutt'.

Available search tags are:
    header subject from to cc bcc list-id
Without a tag, the message headers and body are all searched.

To rebuild the index from scratch (e.g., in case of an aborted
mailpie-add):
    mailpie-index

For more information on commandline options, see mailpie-xxx --help.


TIPS

Because swish-e can take a long time to merge two index files together,
mailpie uses a two-level index system.  mailpie-add puts new messages in
"index.recent".  When this index grows large enough that the merge operation
becomes slow, run "mailpie-index -im" to merge "index.recent" into the main
index (this may take quite a long time).  After that, mailpie-add will be fast
again.


PRINCIPLE

mailpie-add splits mailboxes into individual messages, which are stored
in separate files according to their sha1 hash.

When indexing, mailpie-add and mailpie-index convert each message into
an xml document with markup that indicates headers that should be
indexed specially (e.g., with the <from></from> tag).  These xml
documents are then handed off to swish-e to index.

When searching, mailpie-search involkes swish-e to find the messages
that match the given search terms.  It concatentates each message into
an mbox and then optionally invokes a mailreader.  Optionally,
mailpie-search follows in-reply-to and references to find other messages
in the same thread as matching messages.


EFFICIENCY

When adding thousands of messages, the average rate on a 1.8GHz machine
is 60/second.

When searching tens of thousands of messages, the time is well under 1
second for 100 results when the swish indexes are in memory.


NON-FEATURES

The following features are outside the scope of mailpie and are unlikely
to be added:

 * A Graphical User Interface

 * Support for mailbox formats other than 'mbox'


BUGS

 * There are no known bugs at this time


RELEASE HISTORY

v0.4:
 * mailpie-search: Improved error reporting when swish is unhappy with search
   terms
 * mailpie-search: Improved progress reporting
 * mailpie-search: Mailreader may come from environment variable
 * mailpie-add, mailpie-index: clean up progress messages from swish when
   output is going to email or other non-terminal destination
 * mailpie-admin: new tool to get database statistics or remove messages
   (e.g., spam that got in the index)
 * use a separate threading database.  When upgrading, run mailpie-index
   -f to create this index
 * Documentation in the doc/ directory.  builds using asciidoc or readable
   as-is

v0.3:
 * mailpie-add: fix typo that caused it to error at the end

v0.2:
 * mailpie-add: improve speed of by introducing two levels of index
 * mailpie-add: fix time report when there are zero tens-of-seconds
 * mailpie-search: -l, --after=, --before=, and --limit=  to take arguments
 * mailpie-search: improve -N flag to configure mailreader
 * mailpie-index: add --merge action
 * all: improve usage messages
 * all: improve progress reporting

v0.1:
 * Initial release

COPYRIGHT

Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
                                                                          
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
                                                                          
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


debian/changelog
mailpie (0.4.10) stable; urgency=medium

  * Fix thread database path in search

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 09 May 2016 19:22:06 -0500

mailpie (0.4.9) stable; urgency=medium

  * Add missing import of errno in mailpie-index

 -- Jeff Epler <jepler@unpythonic.net>  Fri, 01 Apr 2016 06:48:15 -0500

mailpie (0.4.8) stable; urgency=medium

  * Change index paths so they can be on their own fs

 -- Jeff Epler <jepler@unpythonic.net>  Sun, 13 Mar 2016 11:16:39 -0500

mailpie (0.4.7) stable; urgency=low

  * Fix mailpie-search -t

 -- Jeff Epler <jepler@unpythonic.net>  Tue, 25 Mar 2014 18:37:38 -0500

mailpie (0.4.6) stable; urgency=low

  * Update to new version

 -- Jeff Epler <jepler@unpythonic.net>  Sat, 22 Mar 2014 12:20:16 -0500

mailpie (0.4.5) lucid; urgency=low

  * mailpie-search: Stop mutt checking a bunch of mailboxes when searching
  * mailpie-search: Fix quoting of message IDs in gmane URLs
  * mkdist: don't use "git-foo" style of invoking git commits

 -- Jeff Epler <jepler@unpythonic.net>  Thu, 18 Jul 2013 09:25:09 -0500

mailpie (0.4.4) lucid; urgency=low

  * Various bugfixes

 -- Jeff Epler <jepler@unpythonic.net>  Sun, 24 Feb 2013 11:31:07 -0600

mailpie (0.4.3) lucid; urgency=low

  * fix installation on 10.04

 -- Jeff Epler <jepler@unpythonic.net>  Tue, 27 Jul 2010 08:01:39 -0500

mailpie (0.4.2~1) hardy; urgency=low

  * Fix things
  * some mailers write invalid things in the in-reply-to header which loused
    up the thread database.  Try harder to accept only plausible-looking
    message IDs in i-r-t and references.

 -- Jeff Epler <jepler@unpythonic.net>  Thu, 02 Apr 2009 07:10:48 -0500

mailpie (0.4.2~0) unstable; urgency=low

  * Initial release.

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 26 May 2008 21:49:30 -0500


mkdist.py
#!/usr/bin/python
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import commands
import gzip
import os
import re
import sys
import tarfile
import time
import StringIO

version_tag = re.compile("^v\d+(\.\d+)*$")

def tag_parse(t):
    return [int(ti) for ti in t[1:].split(".")]

def tag_pad(s, l):
    return s + [0] * (l - len(s))

def tag_sort(a, b):
    a = tag_parse(a)
    b = tag_parse(b)

    a = tag_pad(a, len(b))
    b = tag_pad(b, len(a))

    return cmp(a, b)

def highest_version():
    tags = os.popen("git tag").readlines()
    tags = [t.strip() for t in tags if version_tag.match(t)]
    tags.sort(cmp=tag_sort)
    return tags[-1]

def main(args):
    if len(args) == 1:
        version = args[0]
    elif len(args) > 1:
        raise SystemExit, "Usage: %s [version]" % sys.argv[0]
    else:
        status, gitversion = commands.getstatusoutput("git describe --tags")
        version = highest_version()
        if status != 0 or version != gitversion:
            raise SystemExit, """\
Highest version %r doesn't match description %r.
Specify version number explicitly if this is what you want""" % (
                    version, gitversion)

    version = version.lstrip("v")

    DIRNAME = "%(p)s-%(v)s" % {'p': 'mailpie', 'v': version}
    TARNAME = DIRNAME + '.tar.gz'

    verstream = StringIO.StringIO("%s\n" % version)
    verinfo = tarfile.TarInfo(DIRNAME + "/VERSION")
    verinfo.mode = 0660
    verinfo.size = len(verstream.getvalue())
    verinfo.mtime = time.time()

    tardata = os.popen("git archive --prefix=%(p)s/ v%(v)s"
                            % {'p': DIRNAME, 'v': version}).read()
    tarstream = StringIO.StringIO(tardata)

    tar = tarfile.TarFile(mode="a", fileobj=tarstream)
    tar.addfile(verinfo, verstream)
    tar.close()

    out = gzip.open("../" + TARNAME, "wb")
    out.write(tarstream.getvalue())
    out.close()
    os.system("ls -l ../%s" % TARNAME)

if __name__ == '__main__':
    main(sys.argv[1:])


lib/mailpie/log.py
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import atexit
import fcntl
import os
import struct
import sys
import termios

screen_width = screen_height = None
def screen_size():
    if not os.isatty(2): return 0, 0
    import fcntl
    res = fcntl.ioctl(2, termios.TIOCGWINSZ, "\0" * 4)
    return struct.unpack("hh", res)
screen_height, screen_width = screen_size()

last_width = 0
def progress(message, *args):
    if args: message = message % args
    global last_width
    if screen_width == 0: return
    message = message[:screen_width - 1]
    width = len(message)
    if width < last_width:
        message += " " * (last_width - width)
    sys.stderr.write(message + "\r")
    sys.stderr.flush()
    last_width = width        

def log(message, *args):
    if args: message = message % args
    progress_clear()
    sys.stderr.write(message + "\n");
    sys.stderr.flush()
    
def progress_clear():
    if last_width: progress("")

atexit.register(progress_clear)


lib/mailpie/__init__.py
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__all__ = ['swishindex', 'parsedate', 'log', 'threaddb']


lib/mailpie/threaddb.py
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import bsddb
import os
import re
import mailpie.log

msgid_pat = re.compile("<[^>]*>")
class ThreadDB:
    def __init__(self, base, full):
        db = base + "-volatile/thread.db"
        if full and os.path.exists(db):
            os.unlink(db)
            mailpie.log.log("removed old thread db")
        self.db = bsddb.hashopen(db)
        self._cache = {}

    def close(self):
        for k, v in self._cache.items():
            self.db[k] = " ".join(v)
        self.db.close()

    def threads(self, message):
        h = "".join(message.get_all("in-reply-to", []))
        h = h.replace(" ", "")
        h = h.replace("\r", "")
        h = h.replace("\n", "")
        h = h.replace("\t", "")
        for word in msgid_pat.findall(h): yield word
        h = "".join(message.get_all("references", []))
        h = h.replace(" ", "")
        h = h.replace("\r", "")
        h = h.replace("\n", "")
        h = h.replace("\t", "")
        for word in msgid_pat.findall(h): yield word

    def do_one(self, message, hash):
        msgid = message.get("message-id")
        if msgid is None: return
        if not msgid_pat.match(msgid): return
        self.db["h:" + hash] = msgid
        self.db['m:' + msgid] = hash
        threads = set(self.threads(message))
        self.cache("t:" + msgid).update(threads)
        for t in threads: self.cache("t:" + t).add(msgid)

    def cache(self, k):
        if not k in self._cache:
            if len(self._cache) > 1000:
                while len(self._cache) > 500:
                    k, v = self._cache.popitem()
                    self.db[k] = " ".join(v)
            self._cache[k] = set(self.db.get(k, "").split())
        return self._cache[k]


lib/mailpie/parsedate.py
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import subprocess

def getoutput(cmdline):
    p = subprocess.Popen(cmdline, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
    p.stdin.close()
    return p.stdout.read()

def parsedate(s):
    return int(getoutput(["date", "-d%s" % s, "+%s"]))


lib/mailpie/swishfilter.py
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import email.Message
import email.Parser
import errno
import os
import string
import subprocess
import sys
import tempfile
import time
import rfc822
import mailpie.log
import mailpie.threaddb
import binascii

swishconfig_template = """
DefaultContents XML*
IndexFile %s
MetaNames header subject from to cc bcc list-id date
PropertyNamesDate date
MaxWordLimit 14
"""

parser = email.Parser.Parser()

def _get_index_files(base):
    return [base + "-volatile/index", base + "-volatile/index.recent"]

def get_index_files(base, all=False):
    return [path for path in _get_index_files(base) if all or os.path.exists(path)]

def index_exists(name):
    return os.path.exists(name)

def rename_index(old, new):
    print "rename_index", old, new
    os.rename(old, new)
    os.rename(old + ".prop", new + ".prop")

def unlink_index(name):
    os.unlink(name)
    os.unlink(name + ".prop")

def merge_index(target, additional):
    exists = os.path.exists(target)
    print "merge_index", target, additional, exists
    if not exists and len(additional) == 1:
        rename_index(additional[0], target)
        return

    merge = target + ".merge"
    swish_args = ['mailpie-flog', 'swish-e', '-M']
    if exists: swish_args += [target]
    swish_args += additional + [merge]
    os.spawnvp(os.P_WAIT, swish_args[0], swish_args)
    rename_index(merge, target)
    for index in additional: unlink_index(index)

class LockError(RuntimeError): pass

class Lock:
    def __init__(self, path, delay=.1, max_delay=3600):
        self.path = None
        while 1:
            try:
                fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL)
                break
            except os.error, detail:
                if detail.errno != errno.EEXIST: raise
            if delay > max_delay: raise LockError, "Did not lock after %f seconds" % max_delay
            time.sleep(delay)
            delay = delay * 2
        self.path = path
        os.close(fd)

    def __del__(self):
        if self.path is not None:
            os.unlink(self.path)
            self.path = None

def write_one(path, mtime, data, dest=sys.stdout):
    dest.write("Path-name: %s\nLast-Mtime: %d\nContent-Length: %d\n\n" %(path, int(mtime), len(data)))
    dest.write(data)

null = string.maketrans("", "")
remove = "".join(chr(c) for c in range(32))
hremove = "".join(c for c in (chr(c) for c in range(256)) if not c in string.lowercase + string.uppercase + string.digits + "-")


def escape(s):
    s = s.translate(null, remove)
    s = s.replace("&", "&amp;");
    s = s.replace("<", "&lt;");
    s = s.replace(">", "&gt;");
    return s

def recode(payload, encoding):
    payload = payload.replace("", "")
    if isinstance(payload, unicode):
        return payload.encode("utf-8")
    encoding = encoding or "ascii"
    encoding = encoding.replace(" ", "")
    try:
        payload = payload.decode(encoding)
    except (UnicodeDecodeError, LookupError, binascii.Error):
        try:
            payload = payload.decode("utf-8")
        except UnicodeDecodeError:
            payload = payload.decode("latin-1")
    return payload.encode("utf-8")
 
def recode_header(s):
    if isinstance(s, basestring):
        try:
            s = s.decode("utf-8")
        except UnicodeDecodeError:
            s = s.decode("latin-1")
    return s.encode("utf-8")

def get_payload(m):
    payload = m.get_payload()
    encoding = m.get_content_charset()
    if payload is None:
        mailpie.log.log("Payload is None in %s [%s]", m, repr(m._payload)[:20])
        return ''
    if isinstance(payload, basestring):
        if m.get_content_maintype() not in ("text", None): return ""
        payload = m.get_payload(decode=True) or payload
        payload = recode(payload, encoding)
        return escape(payload)
    result = []
    for p in payload:
        if not isinstance(p, email.Message.Message):
            mailpie.log.log("Unexpected payload item %s", p.__class__.__name__)
            continue
        result.append("<attachment>")
        result.append(get_payload(p))
        result.append("</attachment>")
    return "".join(result)

def parse_date(s):
    try:
        return str(int(rfc822.mktime_tz(rfc822.parsedate_tz(s))))
    except:
        mailpie.log.log("parse_date failed: %s", repr(s)[:50])
        return str(0)

date_fields = set(['date'])

def do_one(since, filename, key, thread_db, data=None, dest=sys.stdout):
    mtime = os.stat(filename).st_mtime
    if since:
        if mtime < since: return False
    if data:
        m = parser.parsestr(data)
    else:
        m = parser.parse(open(filename))
    result = ['<?xml version="1.0" encoding="UTF-8"?><message><header>']
    for h in m.keys():
        h = h.lower()
        for v in m.get_all(h):
            v = v.replace("\n", " ")
            v = v.replace("\t", " ")
            if h in date_fields: v = parse_date(v)
            v = recode_header(v)
            h = h.translate(null, hremove)
            result.append('<%s>%s</%s>\n' % (h, escape(v), h))
    result.append('</header><body>')
    result.append(get_payload(m))
    result.append('</body></message>')
    write_one(key, mtime, "".join(result), dest)

    thread_db.do_one(m, key)

    return True

class Incremental: pass

class Swish:
    def __init__(self, base, since=0, start=None):
        self.base = base
        if since is Incremental:
            lastfile = self.lastfile()
            if os.path.exists(lastfile):
                since = float(open(lastfile).read())
        full = (since == 0)
        self.since = since
        self.start = start
        self.lock = Lock(self.subfile("lock"))
        self.swishconfig = tempfile.NamedTemporaryFile()
        self.swishconfig.write(swishconfig_template % self.subfile("index"))
        self.swishconfig.flush()
        swishargs = ['mailpie-flog', 'swish-e', '-S', 'prog', '-i', 'stdin', '-c', self.swishconfig.name]
        if self.since: swishargs.extend(['-f', self.subfile("incremental")])
        self.swish = subprocess.Popen(swishargs, stdin=subprocess.PIPE)
        self.no_op = True
        self.thread_db = mailpie.threaddb.ThreadDB(self.base, full)

    def subfile(self, ext): return self.base + "-volatile/" + ext
    def lastfile(self): return self.subfile("last")

    def close(self):
        self.thread_db.close()
        self.swish.stdin.close()
        result = self.swish.wait()
        if not self.no_op:
            if result != 0: raise RuntimeError, "swish-e exited with code %d" % result
            if self.since:
                merge_index(self.subfile("index.recent"), [self.subfile("incremental")])
        if self.start:
            open(self.lastfile(), "w").write(str(self.start))

    def merge(self):
        if index_exists(self.subfile("index.recent")):
            merge_index(self.subfile("index"), [self.subfile("index.recent")])

    def do_one(self, filename, key, data=None):
        if do_one(self.since, filename, key, self.thread_db, data, self.swish.stdin):
            self.no_op = False
